lib/thredded/topics_search.rb

```diff
@@ -0,0 +1,58 @@
+require 'thredded/search_parser'
+module Thredded
+  class TopicsSearch
+    def initialize(query, scope)
+      @terms = SearchParser.new(query).parse
+      @scope = scope
+
+      @search_categories = @search_users = @search_text = nil
+    end
+
+    # @return [ActiveRecord::Relation<Thredded::Topic>]
+    def search
+      @scope = @scope.joins(:topic_categories).where(category_id: categories) if categories.present?
+      if text.present? || users.present?
+        [search_topics, search_posts].compact.reduce(:union_all)
+      else
+        @scope
+      end
+    end
+
+    protected
+
+    def search_topics
+      DbTextSearch::FullText.new(@scope, :title).search(text) if text.present?
+    end
+
+    def search_posts
+      posts_scope = Thredded::Post
+      posts_scope = posts_scope.where(user_id: users) if users.present?
+      posts_scope = DbTextSearch::FullText.new(posts_scope, :content).search(text) if text.present?
+      @scope.joins(:posts).merge(posts_scope)
+    end
+
+    def categories
+      @search_categories ||=
+        if @terms['in'].present?
+          DbTextSearch::CaseInsensitive.new(Category, :name)
+            .in(@terms['in']).pluck(:id)
+        else
+          []
+        end
+    end
+
+    def users
+      @search_users ||=
+        if @terms['by']
+          DbTextSearch::CaseInsensitive.new(Thredded.user_class, Thredded.user_name_column)
+            .in(@terms['by']).pluck(:id)
+        else
+          []
+        end
+    end
+
+    def text
+      @terms['text']
+    end
+  end
+end
```

Comment:
Love the functionality here, _however_ ... it could use some test coverage -- more-so as a resource for documentation and exhibiting behavior and less so for correctness or "cover your ass code coverage".  Search method has several branches (looks like 2^2?) so it would be helpful to see some scenarios outlined.

THAT BEING SAID ... I'm more than happy leaving it up to you to merge if you want to add a `# TODO` in here for further work. I obviously trust the code :).